### PR TITLE
Fix ssh deploy and user key constraints (#1357)

### LINKED ISSUE #1357: Deploy Key can't be deleted then reused as Profile key
- Gitea version (or commit ref): Gitea Version: 1.1.0+1-g1d65291
- Git version:  git 1:2.7.4-0ubuntu1
- Operating system: Ubuntu 16.04.2 LTS
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) https://try.gitea.io/cdslashetc/deploy-key-test
  - [ ] No
  - [ ] Not relevant
- Log gist:
I was able to recreate on try.gitea by doing the following:
create private repository
add Deploy Key to repository
delete Deploy Key
add same key to user Profile
attempt to clone repository using the private key

## Description
A user added a key as a Deploy Key not realizing it would be read-only. So, he removed the Deploy Key from the repository then added the same key to his user Profile instead since he is trying to use the Jenkins plugin Git Publisher to create a Tag as a post-build action, which needs to merge.

It is possible that he removed the Deploy Key after he had already added it to his user Profile. So, this problem may be related to #938 

I told the user to create a new key and that of course worked fine.

In serv.log Gitea still seems to think it's the same Deploy Key rather than a normal user key:

2017/03/21 11:32:40 [...io/gitea/cmd/serv.go:216 runServ()] [F] Deploy key access denied: [key_id: 4, repo_id: 48]
2017/03/21 11:38:26 [...io/gitea/cmd/serv.go:216 runServ()] [F] Deploy key access denied: [key_id: 4, repo_id: 61]

I'd say something is going awry in the process to delete the Deploy Key. I'm not using my normal workstation today, trying to get a database tool installed and working so I can dig around in the database.

It could be sufficient just to prevent users from reusing a Deploy Key, it's a bad idea anyway for security reasons.

## Screenshots

**If this issue involves the Web Interface, please include a screenshot**

### PR DESCRIPTION
**Pull request by zeripath**
My understanding is that the intention is:

1. A key can either be an ssh user key or a deploy key. It cannot be both.
2. If a key is a user key - it can only be associated with one user.
3. If a key is a deploy key - it can be used in multiple repositories and the permissions it has on those repositories can be different.
4. If a repository is deleted, its deploy keys must be deleted too.

We currently don't enforce any of this and multiple repositories access with different permissions doesn't work at all. This PR enforces the following constraints:

- [x] You should not be able to add the same user key as another user
- [x] You should not be able to add a ssh user key which is being used as a deploy key
- [x] You should not be able to add a ssh deploy key which is being used as a user key
- [x] If you add an ssh deploy key to another repository you should be able to use it in different modes without losing the ability to use it in the other mode.
- [x] If you delete a repository you must delete all its deploy keys.

Fix #1357

### CODE REVIEW

File: cmd/serv.go [hunk 1]
```diff
@@ -233,19 +233,20 @@ func runServ(c *cli.Context) error {
...
 				fail("Key access denied", "Failed to access internal api: [key_id: %d, repo_id: %d]", key.ID, repo.ID)
 			}
-			if !has {
+
+			if deployKey == nil {
 				fail("Key access denied", "Deploy key access denied: [key_id: %d, repo_id: %d]", key.ID, repo.ID)
 			}
 
+			if deployKey.Mode < requestedMode {
+				fail("Key permission denied", "Cannot push with deployment key: %d to repo_id: %d", key.ID, repo.ID)
```
Comment (techknowlogick) [inline cmd/serv.go]: One tiny nit: I know this was the error message before your change, but could you update it to include something about this specific deploy key being read-only?

**Review by techknowlogick — APPROVED** [review]
Looks great! One nit, but I'm ok if this PR remains same.

**Comment by lunny** [comment]
The deploykey should be deleted from public_key table if last repo which use it is deleted?

**Comment by zeripath** [comment]
Hi @lunny the call to deleteDeployKey manages the deletion of the PublicKey entry as necessary.

**Comment by zeripath** [comment]
OK I've added some tests and changed the failure message as per @techknowlogick 

If you look at the tests - I've created them in a more declarative style similar to that of TestGit.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
please backport to release/v1.7

**Pull request merged**